bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java

```diff
@@ -12,34 +12,87 @@
  */
 package org.openhab.core.ui.internal.components;
 
-import java.util.HashMap;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
-import org.openhab.core.storage.StorageService;
+import org.eclipse.jdt.annotation.NonNull;
+import org.openhab.core.common.registry.ManagedProvider;
+import org.openhab.core.ui.components.RootUIComponent;
 import org.openhab.core.ui.components.UIComponentRegistryFactory;
 import org.osgi.service.component.annotations.Component;
 import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
 
 /**
- * Implementation for a {@link UIComponentRegistryFactory} using a {@link StorageService} and a
- * {@link UIComponentProvider}.
+ * Implementation for a {@link UIComponentRegistryFactory} using a set of {@link UIComponentProvider}.
  *
  * @author Yannick Schaus - Initial contribution
+ * @author Łukasz Dywicki - Removed explicit dependency on storage providers.
+ * @author Jonathan Gilbert - Made providers' collections immutable.
  */
 @Component(service = UIComponentRegistryFactory.class, immediate = true)
 public class UIComponentRegistryFactoryImpl implements UIComponentRegistryFactory {
-    Map<String, UIComponentRegistryImpl> registries = new HashMap<>();
-
-    @Reference
-    StorageService storageService;
+    Map<String, UIComponentRegistryImpl> registries = new ConcurrentHashMap<>();
+    Map<String, Set<UIProvider>> providers = new ConcurrentHashMap<>();
 
     @Override
     public UIComponentRegistryImpl getRegistry(String namespace) {
         UIComponentRegistryImpl registry = registries.get(namespace);
         if (registry == null) {
-            registry = new UIComponentRegistryImpl(namespace, storageService);
+            Set<UIProvider> namespaceProviders = this.providers.get(namespace);
+            ManagedProvider<@NonNull RootUIComponent, @NonNull String> managedProvider = null;
+            if (namespaceProviders != null) {
+                for (UIProvider provider : namespaceProviders) {
+                    if (provider instanceof ManagedProvider) {
+                        managedProvider = (ManagedProvider<@NonNull RootUIComponent, @NonNull String>) provider;
```

Comment:
Yes they are, because otherwise the null constraints are not fulfilled. The reason is that the whole class is not null-annotated. Since the other issues are only typos / year correction and we are allowed to edit the PR, are you ok with merging after the suggestions have been committed?

Reply:
OK I've added the annotation on the class and done some more trivial cleanup/fixes.
LGTM now, so if you are happy with it as well, feel free to merge it. :slightly_smiling_face: